server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -151,6 +154,13 @@ public StoredContext stashContext() {
             threadContextStruct = threadContextStruct.putTransient(TASK_ID, context.transientHeaders.get(TASK_ID));
         }
 
+        if (context.transientHeaders.containsKey(CURRENT_SPAN)) {
```

Comment:
This is why we introduced `TracingContextPropagator`, right? I would imagine we should be doing that instead:

```
final Span span = tracingTelemetry.getContextPropagator().extract(context.transientHeaders);
```

Reply:
I dont think we are splitting the logic of context propagation between `ThreadContextBasedTracerContextStorage ` and `ThreadContext`. `ThreadContextBasedTracerContextStorage` contains the logic to add/update the current span in the thread context. However the logic currently in `ThreadContext` relating to tracing related key is only to ensure the key(and its value) is copied to new ThreadContext(transient header to be precise) when it is recreated during stashing or copying

Reply:
Sure

Reply:
The issue become more or less clear, so we have a cycle between Tracer <-> ThreadContext (through `ThreadContextBasedTracerContextStorage` and static `TracerManager`), looking how to break it apart

Reply:
Ok, here is what we could do:
 - introduce new interface `ThreadContextStateExtractor`:
```
public interface ThreadContextStateExtractor {
    Map<String, Object> transients(Map<String, Object> source);
}
```

 - add to `ThreadContext`:
      `private final List<ThreadContextStateExtractor> extractors = new CopyOnWriteArrayList<>();`
       methods     `addStateExtractor` / `removeStateExtractor`

- use it like this (we could optimize `threadContextStruct` to accept many headers in `putTransient`, not only one)

```
        final Map<String, Object> transientHeaders = extractors().exract(context.transientHeaders);
        if (!transientHeaders.isEmpty()) {
            threadContextStruct = threadContextStruct.putTransient(transientHeaders);
        }

```
 - implement `ThreadContextStateExtractor` by `ThreadContextBasedTracerContextStorage`, since it accepts `threadContext`, it could register itself by adding `threadContext.addStateExtractor(this)`

With this model, we should be able to remove TASK_ID & other contextual manipulations from `ThreadContext` class (could be done outside of this pull request).

Reply:
Thanks for the suggestion, this looks good. Will address this in follow-up PR